sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityCheckerSuite.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming.state
+
+import java.util.UUID
+
+import scala.util.Random
+
+import org.apache.hadoop.conf.Configuration
+
+import org.apache.spark.sql.execution.streaming.state.StateStoreTestsHelper.newDir
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types._
+
+class StateSchemaCompatibilityCheckerSuite extends SharedSparkSession {
+
+  testQuietly("changing schema of state when restarting query") {
```

Comment:
What is your reasoning for putting all of these checks into a single test? It seems like this should be several independent tests so that it is easy to see which cases are broken simultaneously (for example, type change and nullability change). Putting tests in one big case means that failing tests prevent other tests from running.

Reply:
I don't have a strong reasoning about that; while I see it hasn't been consistent across tests regarding the test coverage of single unit test, I generally agree it's good to have separate tests if that doesn't bring too much overhead (code dup or something). Let me check this again.